### PR TITLE
fix: close should gracefully wait for all queued requests

### PR DESCRIPTION
**Pull request by ronag**
Fixes: https://github.com/mcollina/undici/issues/20

### CODE REVIEW

File: lib/client.js [hunk 1]
```diff
@@ -402,8 +386,6 @@ class Client extends EventEmitter {
     }
 
     if (this.socket) {
-      // TODO make sure we error everything that
-      // is in flight
```
Comment (ronag) [inline lib/client.js:406]: This was fixed in one of the other PR's. Just removing the TODO.

File: lib/client.js [hunk 2]
```diff
@@ -375,18 +370,7 @@ class Client extends EventEmitter {
...
-    // TODO test this
-    if (
-      this[kQueue].length() === 0 &&
-      this[kCallbacks].length === 0 &&
-      this.socket
-    ) {
-      this.socket.end()
-      // TODO wait util socket is closed before setting to null?
-      this.socket = null
-    }
```
Comment (ronag) [inline lib/client.js:389]: Don't end the socket if _lastBody exists. Could use allowHalfOpen but there is not much advantage to that, just destroy the socket once we are done.

File: lib/client.js [hunk 3]
```diff
@@ -359,13 +359,8 @@ class Client extends EventEmitter {
     }
     this.closed = true
 
-    // TODO: in-flight should still be able to timeout?
-    if (this[kTimer]) {
-      this[kTimer].clear()
-      this[kTimer] = null
-    }
```
Comment (ronag) [inline lib/client.js:366]: close waits for requests so timeout is still relevant

**Comment by ronag** [comment]
flaky test

File: test/client-errors.js [hunk 4]
```diff
@@ -27,7 +27,7 @@ test('GET errors and reconnect with pipelining 1', (t) => {
     const client = new Client(`http://localhost:${server.address().port}`, {
       pipelining: 1
     })
-    t.tearDown(client.close.bind(client))
+    t.tearDown(client.destroy.bind(client))
```
Comment (ronag) [inline test/client-errors.js:30]: This tests don't actually finish the requests which means if we do close we have to wait for a timeout... just do destroy instead

File: lib/client.js [hunk 5]
```diff
@@ -354,38 +354,33 @@ class Client extends EventEmitter {
...
+        })
+      })
     }
+
     this.closed = true
 
-    // TODO: in-flight should still be able to timeout?
-    if (this[kTimer]) {
-      this[kTimer].clear()
-      this[kTimer] = null
```
Comment (ronag) [inline lib/client.js:365]: close waits for requests so timeout is still relevant

File: lib/client.js [hunk 6]
```diff
@@ -354,38 +354,33 @@ class Client extends EventEmitter {
...
   close (cb) {
-    if (this.closed) {
-      return
+    if (cb === undefined) {
+      return new Promise((resolve, reject) => {
+        this.close((err, data) => {
+          return err ? reject(err) : resolve(data)
+        })
+      })
     }
```
Comment (ronag) [inline lib/client.js:363]: promisify

File: lib/client.js [hunk 7]
```diff
@@ -354,38 +354,33 @@ class Client extends EventEmitter {
   }
 
   close (cb) {
-    if (this.closed) {
-      return
```
Comment (mcollina) [inline lib/client.js:358]: I think those two lines might still be needed somewhere?

Comment (ronag) [reply lib/client.js:358]: No, not required. If it's already closed it doesn't do anything.

**Pull request merged**